src/iter/mod.rs

```diff
@@ -38,61 +38,45 @@ impl<T: BitSetLike> BitIter<T> {
     }
 }
 
+enum State<T> {
+    Empty,
+    Continue,
+    Value(T)
+}
+
 impl<T> Iterator for BitIter<T>
     where T: BitSetLike
 {
     type Item = Index;
 
     fn next(&mut self) -> Option<Self::Item> {
-        loop {
-            // Look at first level
-            if self.masks[0] != 0 {
-                // Take first bit that isn't zero
-                let bit = self.masks[0].trailing_zeros();
-                // Remove it from masks
-                self.masks[0] &= !(1 << bit);
-                // and returns it's index
-                return Some(self.prefix[0] | bit);
-            }
-            // Look at second level
-            if self.masks[1] != 0 {
-                // Take first bit that isn't zero
-                let bit = self.masks[1].trailing_zeros();
-                // Remove it from masks
-                self.masks[1] &= !(1 << bit);
-                // Calculate index of the bit in first level
-                let idx = self.prefix[1] | bit;
-                // Take corresponding usize from layer below
-                self.masks[0] = self.set.layer0(idx as usize);
-                // Prefix of the complete index
-                self.prefix[0] = idx << BITS;
-                continue;
-            }
-            // Look at third level
-            if self.masks[2] != 0 {
-                // Take first bit that isn't zero
-                let bit = self.masks[2].trailing_zeros();
-                // Remove it from masks
-                self.masks[2] &= !(1 << bit);
-                // Calculate index of the bit in second level
-                let idx = self.prefix[2] | bit;
-                // Take corresponding usize from layer below
-                self.masks[1] = self.set.layer1(idx as usize);
-                // Prefix of the index of the second level
-                self.prefix[1] = idx << BITS;
-                continue;
+        use self::State::*;
+        let mut handle_level = |level: usize| if self.masks[level] == 0 {
```

Comment:
I thought that I made this comment before already, but somehow it didn't work. Please make this a method on the struct; such a long closure makes the source code harder to understand and read.